Reverse ledger chain verification

marta-lokhova opened the pull request:
In this PR, ledger chain verification is updated to verify checkpoint starting from the recent one and working its way back to the beginning of the range. Additionally, ledger chain verification work does not make calls to ledger manager anymore, as the trusted SCP hash is passed into CatchupWork directly.

vogel commented on src/catchup/CatchupWork.cpp:
```diff
@@ -25,15 +25,15 @@ namespace stellar
-                         bool manualCatchup, ProgressHandler progressHandler,
-                         size_t maxRetries)
+                         ProgressHandler progressHandler, size_t maxRetries,
+                         optional<Hash> trustedScpHash)
```
maxRetries is last argument for constructors of all Work items, could you please leave it last here as well for consistency?

vogel commented on src/history/HistoryTests.cpp:
```diff
@@ -554,7 +555,7 @@ TEST_CASE("Catchup recent", "[history][catchuprecent]")
     for (auto a : apps)
     {
-        catchupSimulation.catchupApplication(initLedger, 80, false, a);
+        REQUIRE(catchupSimulation.catchupApplication(initLedger, 80, false, a));
```
nice catch!

vogel commented on src/history/HistoryTests.cpp:
```diff
@@ -609,11 +610,14 @@ TEST_CASE("Catchup manual", "[history][catchupmanual]")
-                    initLedger1, std::numeric_limits<uint32_t>::max(), true, a);
+                REQUIRE(catchupSimulation.catchupApplication(
+                    initLedger1, std::numeric_limits<uint32_t>::max(), true,
+                    a));
```
I think 'a' variable is not longer necessary here. You can just do REQUIRE on line 610 and use plain 'true' on 617 and 620.

vogel replied:
ok, for some reason I was expecting this to be bool.

vogel commented on src/catchup/VerifyLedgerChainWork.cpp:
```diff
@@ -94,46 +95,41 @@ VerifyLedgerChainWork::getStatus() const
-    }
-    if (mFirstVerified.header.ledgerSeq != 0)
-    {
-        mFirstVerified = setLedger;
```
Is it save to not reset mFirstVerified here? I think so, as mFirstVerified is verified at the end of this work, so the work cannot fail after that. If my analysis is correct, we should have a comment here explaining that.

vogel commented on src/catchup/VerifyLedgerChainWork.cpp:
```diff
@@ -197,56 +207,91 @@ VerifyLedgerChainWork::verifyHistoryOfSingleCheckpoint()
     {
-        throw std::runtime_error("Verification overshot target ledger");
+        throw std::runtime_error(
+            "Verification overshot first ledger in the range.");
```
overshot -> undershot

MonsieurNicolas commented on src/catchup/CatchupWork.cpp:
```diff
@@ -172,8 +174,9 @@ CatchupWork::verifyLedgers(LedgerRange const& range)
-    mVerifyLedgersWork = addWork<VerifyLedgerChainWork>(
-        *mDownloadDir, range, mManualCatchup, mFirstVerified, mLastVerified);
+    mVerifyLedgersWork =
+        addWork<VerifyLedgerChainWork>(*mDownloadDir, range, mFirstVerified,
```
we should probably use this occasion to get rid of this "pass by reference" that is quite ugly (`mFirstVerified`) and set it when `mVerifyLedgersWork` succeeds with something like `mFirstVerified  = mVerifyLedgersWork->getFirstVerified()`.
Also, maybe rename it to something a bit less cryptic - I don't know what "first verified" means especially now that we verify backwards. Maybe `mVerifiedRangeStart` ? (also rename the corresponding member in `VerifyLedgerChainWork`)

MonsieurNicolas commented on src/catchup/VerifyLedgerChainWork.cpp:
```diff
@@ -94,46 +95,48 @@ VerifyLedgerChainWork::getStatus() const
-        mFirstVerified = setLedger;
-    }
-    if (mLastVerified.header.ledgerSeq != 0)
+    // Note that mFirstVerified is not reset here, since it is set
```
Instead of this comment, I think it would be much cleaner to always reset `mFirstVerified` to the zero value.

MonsieurNicolas commented on src/catchup/VerifyLedgerChainWork.cpp:
```diff
@@ -94,46 +95,48 @@ VerifyLedgerChainWork::getStatus() const
+    // Note that mFirstVerified is not reset here, since it is set
+    // in the end of verification, where work wouldn't fail
+
+    if (mVerifiedAhead.header.ledgerSeq != 0)
```
why the condition on resetting `mVerifiedAhead` ?

marta-lokhova replied:
This was in the original code, looks like it's unnecessary, so I removed it.

MonsieurNicolas commented on src/catchup/VerifyLedgerChainWork.cpp:
```diff
@@ -197,56 +229,98 @@ VerifyLedgerChainWork::verifyHistoryOfSingleCheckpoint()
-                              << curr.header.ledgerSeq << " with LedgerManager";
-        status = mApp.getLedgerManager().verifyCatchupCandidate(curr,
-                                                                mManualCatchup);
+        // Verifying the most recent checkpoint means no other checkpoints are
```
I think that we should refactor those two ifs: there is no reason to distinguish the "trusted ledger from SCP" and the "trusted ledger from the next checkpoint":
I would just set `nextCheckpointFirstLedger` to either the trusted value from SCP (if we're verifying the last checkpoint) or to the value of `mVerifiedAhead`

marta-lokhova replied:
Done.

MonsieurNicolas commented on src/catchup/VerifyLedgerChainWork.h:
```diff
@@ -24,28 +24,36 @@ class VerifyLedgerChainWork : public Work
     medida::Meter& mVerifyLedgerSuccess;
     medida::Meter& mVerifyLedgerFailureLedgerVersion;
     medida::Meter& mVerifyLedgerFailureOvershot;
+    medida::Meter& mVerifyLedgerFailureUndershot;
```
I don't think we care to have that kind of granularity for metrics

marta-lokhova replied:
Updated this, and kept "ledger-verify" success and failure metrics only.

MonsieurNicolas commented on src/catchup/VerifyLedgerChainWork.cpp:
```diff
@@ -197,56 +229,98 @@ VerifyLedgerChainWork::verifyHistoryOfSingleCheckpoint()
+    }
+    else
+    {
+        CLOG(INFO, "History")
```
this should be `DEBUG` not `INFO`

MonsieurNicolas commented on src/catchup/VerifyLedgerChainWork.cpp:
```diff
@@ -197,56 +229,98 @@ VerifyLedgerChainWork::verifyHistoryOfSingleCheckpoint()
+    assert(ledger.header.ledgerSeq == mRange.last());
+    if (!mTrustedHash)
+    {
+        CLOG(INFO, "History")
```
this should be `DEBUG` not `INFO`

MonsieurNicolas commented on src/ledger/LedgerManager.h:
```diff
@@ -160,7 +160,8 @@ class LedgerManager
     virtual void startCatchup(CatchupConfiguration configuration,
-                              bool manualCatchup) = 0;
+                              bool manualCatchup,
+                              optional<Hash> trustedHash) = 0;
```
similar comment than for `catchupHistory` regarding `trustedHash`

marta-lokhova commented on src/catchup/CatchupManager.h:
```diff
@@ -30,8 +30,8 @@ class CatchupManager
-                                bool manualCatchup,
-                                CatchupWork::ProgressHandler handler) = 0;
+                                CatchupWork::ProgressHandler handler,
+                                optional<Hash> trustedHash) = 0;
```
Good idea, I think it would definitely make more sense to have ledger hash in `CatchupConfiguration`. Will update.

marta-lokhova commented on src/catchup/CatchupWork.cpp:
```diff
@@ -129,14 +130,15 @@ CatchupWork::hasAnyLedgersToCatchupTo() const
     CLOG(INFO, "History")
         << "Last closed ledger is later than current checkpoint: "
-        << mLastClosedLedgerAtReset << " > " << mRemoteState.currentLedger;
+        << mLastClosedLedgerAtReset.header.ledgerSeq << " > "
```
Is there more information on the bug? (a github issue maybe?)

Regarding retrying with a different archive, yes, that would be good. In order for this to happen, we first need to ensure that a single archive is used during catchup (not a guarantee right now, archive is selected randomly each download).

MonsieurNicolas commented:
needs to be rebased

MonsieurNicolas commented:
I guess there were comments, maybe it's easier to first address comments before rebasing

marta-lokhova commented:
FYI, this is rebased and comments are addressed.

MonsieurNicolas commented:
needs to be rebased again @marta-lokhova

marta-lokhova commented:
Rebased and updated. A few things:
* I added better handling of last closed ledger to the verification logic
* I added some ledger header generation infrastructure for testing as well as a verification test.

MonsieurNicolas commented on src/catchup/VerifyLedgerChainWork.cpp:
```diff
@@ -150,25 +151,35 @@ VerifyLedgerChainWork::verifyHistoryOfSingleCheckpoint()
         }
+        // Verify LCL that is just before the first ledger in range
+        else if (curr.header.ledgerSeq == mLastClosed.first + 1 &&
+                 curr.header.ledgerSeq ==
```
I don't think you need the second condition: if `curr.header.ledgerSeq == mLastClosed.first + 1` -> the hashes must agree.

marta-lokhova replied:
Added the second condition to avoid potentially checking the same link twice (if LCL is inside the range, it'd be verified later anyway), but it can be removed since the overhead is likely negligible.

MonsieurNicolas replied:
yeah I agree, it just makes the code harder to read to have this extra condition

MonsieurNicolas commented on src/catchup/VerifyLedgerChainWork.cpp:
```diff
@@ -150,25 +151,35 @@ VerifyLedgerChainWork::verifyHistoryOfSingleCheckpoint()
+            auto lclResult = verifyLedgerHistoryLink(*mLastClosed.second, curr);
+            if (lclResult != HistoryManager::VERIFY_STATUS_OK)
+            {
+                return lclResult;
```
need logging for this code path

marta-lokhova commented:
@MonsieurNicolas, addressed your comments (the PR is squashed and rebased now too)

MonsieurNicolas commented:
r+ 6274c089f7ec335274139250fff310ee4edc97aa
